pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+		if sc.Parameters[driver.ParamMultishareInstanceScLabel] == scTag {
+			return sc, nil
+		}
+	}
+	return nil, fmt.Errorf("no storageclass match storageClassTag %q", scTag)
+}
+
+func (recon *multishareReconciler) assignInstanceInfo(instanceInfo *v1alpha1.InstanceInfo, shareName string) (*v1alpha1.InstanceInfo, error) {
+	instanceInfoClone := instanceInfo.DeepCopy()
+	if instanceInfoClone.Status == nil {
+		instanceInfoClone.Status = &v1alpha1.InstanceInfoStatus{}
+	}
+	for _, name := range instanceInfoClone.Status.ShareNames {
+		if name == shareName {
+			return instanceInfo, nil
+		}
+	}
+	klog.Infof("appending share %q to %q instanceInfo's assigned share list", shareName, instanceInfoClone.Name)
+	instanceInfoClone.Status.ShareNames = append(instanceInfoClone.Status.ShareNames, shareName)
+
+	return recon.updateInstanceInfoStatus(context.TODO(), instanceInfoClone)
+}
+
+func (recon *multishareReconciler) assignShareInfo(shareInfo *v1alpha1.ShareInfo, instanceURI string) (*v1alpha1.ShareInfo, error) {
+	shareInfoClone := shareInfo.DeepCopy()
+	if shareInfoClone.Status == nil {
+		shareInfoClone.Status = &v1alpha1.ShareInfoStatus{}
+	}
+	shareInfoClone.Status.InstanceHandle = instanceURI
+	klog.Infof("try to assign share %q to instance %q", shareInfo.Name, instanceURI)
+	return recon.updateShareInfoStatus(context.TODO(), shareInfoClone)
+}
+
+// createAndUpdateInstanceInfos create instanceInfo objects if needed and updates their statuses to match with reality
+// InstanceInfo objects in the returned map must be treated as read only
+func (recon *multishareReconciler) createAndUpdateInstanceInfos(instances []*file.MultishareInstance, instanceShares map[string][]*file.Share) map[string]*v1alpha1.InstanceInfo {
+	instanceInfoMap := make(map[string]*v1alpha1.InstanceInfo)
+
+	for _, instance := range instances {
+		instanceURI, err := file.GenerateMultishareInstanceURI(instance)
+		if err != nil {
+			klog.Errorf("couldn't generate instanceURI: %v for instance %q", err, instance.Name)
+			continue
+		}
+		iiName := util.InstanceURIToInstanceInfoName(instanceURI)
+
+		instanceInfo, err := recon.instanceLister.Get(iiName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				klog.Errorf("error getting instanceInfo %q from informer: %v", iiName, err)
+				continue
+			}
+			instanceInfo, err = recon.clientset.MultishareV1alpha1().InstanceInfos().Get(context.TODO(), iiName, metav1.GetOptions{})
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					klog.Errorf("error getting instanceInfo %q from api server: %v", iiName, err)
+					continue
+				}
+				klog.V(6).Infof("instanceInfo object for instance %q not found in API server", instance.Name)
+				instanceInfo = nil
+			}
+			klog.V(6).Infof("instanceInfo object for instance %q not found in informer cache but found in API server", instance.Name)
+		}
+
+		instanceInfo, err = recon.reconstructInstanceInfo(iiName, instance, instanceInfo)
```

Comment:
reconstructInstanceInfo() is creating the actual CR. Instead, why cant we do a single update to the kube-api server based on the diff of reconstructInstanceInfo() CR and  maybeUpdateInstanceInfoStatus() ?